### DISCUSSION TITLE
How many threads Node creates per core when using worker threads?

### DISCUSSION THREAD

**Discussion by crealityeducation** [jasnell/piscina#478] [Q&A] [answered]
Dears,

I'm running node js worker threads on MacBook Air M1. There are 8 core. And using PISCINA to create threads. I'm getting 12 threads. Are those for one core or on all 8 core. How do I know if main thread is blocked/being used in worker threads when running workers for all threads?

When running on all threads, other requests are getting blocked. I want to use all cores and full capacity of threads without blocking main thread.

Also, when I see the CPU utilisation when worker threads are running, its not crossing above 30%. It looks like threads are spawned on single core.

 I really need performance here. And I see the CPU utilization is not crossing 25% while in worker threads mode. It looks like I'm underutilizing resources.

**Comment by metcoder95** [comment]
👋 

It depends on your workload. CPU scheduling is tricky; sometimes the workload does not require all the threads to be active nor spawned simultaneously and, for instance, going idle.

If you want to use all your threads, I suggest you to tweak `concurrentTasksPerWorker` based on your workload to enhance the distribution across your pool of threads.

Also remember that the task balancing goes in a `FIFO` mode

**Comment by crealityeducation** [comment]
Hello Carlos, 
     Thanks for the response. So, can I increase the pool of threads or its constant/CPU core? I'm doing CPU intensive numerical work and its taking 7sec (min) - 10 sec (max) using all threads (12). I'm using maxThreads to get the threads and then distributing work.

**Reply by metcoder95** [reply, marked as answer]
Yes, for sure; you can increase it as many as you need to tweak the performance you are seeking for

**Reply by crealityeducation** [reply]
many thanks Carlos for inputs. Have a great day.

**Reply by crealityeducation** [reply]
Hello Carlos,
     Just out of curiosity, is there any logic for 'maxThread= 1.5 * number of core'?

**Reply by metcoder95** [reply]
Not really, each workload behaves diferently and for instance, to reach its best performance, it will require different number of threads.

The default is just a _best effort_ to provide a good starting number to kick-off your application.
